Handle shared memory cases in MathBitFallback

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #63602

This PR fixes the case when a read and write is performed on a memory shared between mutable and (or) non-mutable arguments. Example:
```
a=torch.tensor([1+1j])
b=a.conj()
b.add_(a) # should return tensor([2]) but returns tensor ([2-2j])
```

The issue here is that in the conjugate fallback, we resolve the conjugation in-place for mutable arguments which can be a problem as shown above in the case when other input arguments share memory with the mutable argument(s). 
This PR fixes this issue by:
1. first scanning through the operator input arguments and creating a vector of mutable arguments that have the conj bit set to `True` (and accordingly setting the flag `check_for_alias_with_mut_arg ` to `True` or `False`).
2. Iterating through all the arguments. At this time we only look at the non-mutable arguments. If `check_for_alias_with_mut_arg` is set to `True`, then we iterate through `mutable_inputs` to check if the current arg tensor in question doesn't alias any of the entries in `mutable_inputs`. If yes, then we clone the non-mutable tensor arg, else we resolve the conjugation as before.
3. Now we look through the mutable_inputs vector (which contains only mutable input tensors with conj bit set to `True`). We in-place conjugate each of the entries in the vector.
4. Do the computation.
5. Re-conjugate the mutable argument tensors.

NOTE: `TensorLists` are not fully handled in ConjugateFallback. Please see the in-line comment for more details.

Fixes https://github.com/pytorch/pytorch/issues/59943

Differential Revision: [D30466905](https://our.internmc.facebook.com/intern/diff/D30466905)

cc @ezyang @anjali411 @dylanbespalko @mruberry  @nikitaved

anjali411 committed:
Handle shared memory cases in MathBithFallback

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/MathBitsFallback.h:
```diff
@@ -60,6 +63,18 @@ struct MathOpFallback {
+          auto& ivalue = (*stack)[stack_start + i];
+          //TODO: add a table listing all possible cases and clearly state what works/is allowed
+          if (ivalue.isTensor()) {
+            auto mut_arg_tensor = ivalue.toTensor();
```
this incurs a refcount bump. Use `const auto&`

anjali411 replied:
yup done!

ezyang commented:
PR description?

ezyang commented on aten/src/ATen/native/MathBitsFallback.h:
```diff
@@ -72,55 +87,79 @@ struct MathOpFallback {
+        c.4. shared memory between two or more non-mutable tensor args
+        c.5. shared memory between a tensor and tensorlist arg
+        ...
+        c.n. no shared memory between args
```
It would be nice to be able to prove this implementation correct without having to do a giant case split.

anjali411 replied:
true for now I am just not handling c. because it's too complicated + expensive

ezyang commented on aten/src/ATen/native/MathBitsFallback.h:
```diff
@@ -72,55 +87,79 @@ struct MathOpFallback {
         // Was already tested by is_write loop above
         TORCH_INTERNAL_ASSERT_DEBUG_ONLY(argument.alias_info()->isWrite());
-        mut_arg = true;
+        continue;
```
continue'ing here looks sus. If the argument is mutable AND conjugate bit is set, passing it on directly seems wrong because the kernel is going to ignore the conjugate bit and stomple all over it.

anjali411 replied:
no because we only wanna handle the non-mutable args in this loop and not modify the memory until we verify either the non-mutable args don't share storage with mutable args (with conj bit set to `True`) else clone them. 

Then we loop through the mutable args with conj bit set to `True`, and in-place conjugate them.

ezyang commented:
I got lost in the implementation. High level description will help a lot.

anjali411 committed:
Update on "Handle shared memory cases in MathBithFallback"


Fixes https://github.com/pytorch/pytorch/issues/59943

[ghstack-poisoned]

anjali411 commented:
> I got lost in the implementation. High level description will help a lot.

added PR description explaining roughly what's happening in fallback now (in order). lmk if more clarification is needed.

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30466905).

ezyang commented:
> Iterating through all the arguments. At this time we only look at the non-mutable arguments. If check_for_alias_with_mut_arg is set to True, then we iterate through mutable_inputs to check if the current arg tensor in question doesn't alias any of the entries in mutable_inputs. If yes, then we clone the non-mutable tensor arg, else we resolve the conjugation as before.

I feel like there is a policy decision we need to make here. An alternative to scanning for aliasing is to just directly assume that every argument may alias each other, and not do the in-place mutation trick. So then it will look something like:

```
def add_(a, b):
    if a.is_conj():
        if b.is_conj():
            a.copy_(a.clone().add_(b.clone()).conj())
        else:
            a.copy_(a.clone().add_(b).conj())
    else:
        a.add_(b.clone())
```

The no aliasing case simply saves us a copy of a, but we're already doing a bunch of copying in the fallback, and it comes at the cost of having to do a bit of extra work in the body of the function.

ezyang commented:
@bdhirsh it seems like conjugate fallback would be simplified if it could assume no aliases :)

ezyang commented on aten/src/ATen/native/MathBitsFallback.h:
```diff
@@ -44,6 +44,9 @@ struct MathOpFallback {
     const auto stack_start = stack->size() - num_arguments;
```
The comment above these lines (which gives the general algorithm) should get updated with the new algo

ezyang commented on aten/src/ATen/native/MathBitsFallback.h:
```diff
@@ -44,6 +44,9 @@ struct MathOpFallback {
     c10::optional<bool> is_write;
+    // Mutable inputs to be tracked separately
+    std::vector<Tensor> mutable_inputs;
+    bool check_for_alias_with_mut_arg = false;
```
A good comment here would be an explanation of all the things we precompute. Previously, we just tested if the function overall was a "write" function (though arguably should have explained what `is_write` meaning was); now there's three things we are precomputing and so it is a good idea to just generally say what the semantics of these variables are.

ezyang replied:
In particular, `mutable_inputs` here only include *conjugated* mutable inputs. Variable name here doesn't make this clear at all!

ezyang replied:
Similarly, `check_for_alias_with_mut_arg` doesn't seem necessary at all; it is always true when `mutable_inputs` is non-empty

ezyang commented on aten/src/ATen/native/MathBitsFallback.h:
```diff
@@ -72,55 +87,79 @@ struct MathOpFallback {
+            if (tensor.is_alias_of(mutable_input)) {
+              tensor = tensor.clone();
+              cloned = true;
+              continue;
```
Why is this a continue? Don't you want break?

anjali411 replied:
oh yeah this should be break!

ezyang commented on aten/src/ATen/native/MathBitsFallback.h:
```diff
@@ -72,55 +87,79 @@ struct MathOpFallback {
+          bool cloned = false;
+          for (const auto& mutable_input : mutable_inputs) {
+            if (tensor.is_alias_of(mutable_input)) {
+              tensor = tensor.clone();
```
I have an alternate proposal. Right now, you have:

```
// precondition: self is conjugate, self aliases with other
// what currently happens
other = other.clone()
self.set_bit_(false)
math_op_(self)

redispatch_(self, other)

self.set_bit_(true)
math_op_(self)
```

Here is the proposed alternative:

```
// precondition: self is conjugate (no condition on other)
// other possibly policy
self_clone = self.clone()
redispatch_(self_clone, resolve_conj(other))
self.copy_(self_clone)
```

This is quite a lot simpler, and even better, it doesn't require the aliasing test. It is slightly worse, performance wise, in the case where self is conjugate but other doesn't alias self; in this case, we will do an unnecessary copy of self to clone to do the conjugate operation on it (indeed, this is precisely what the old optimization was about optimizing). But it doesn't seem worth the ensuing complication to detect aliases.

anjali411 replied:
Here's roughly the perf comparison for the two scenarios:
CPU:
```
M=torch.randn(8, dtype=torch.cdouble).conj() # mutable conjugate view tensor
N=torch.randn(8, dtype=torch.cdouble) # f(self_clone, args)

[-------------- Effective materialization of conj --------------]
                          |  Using clone  |  Using conj_physical_
1 threads: ------------------------------------------------------
      M.clone()           |     2385.4    |
      M.conj_physical_()  |               |         737.8

Times are in nanoseconds (ns).

[-------------- Effective materialization of conj --------------]
                          |  Using copy_  |  Using conj_physical_
1 threads: ------------------------------------------------------
      M.copy_(N)          |     921.1     |
      M.conj_physical_()  |               |         620.6

Times are in nanoseconds (ns).
```
CUDA:
```
[-------------- Effective materialization of conj --------------]
                          |  Using clone  |  Using conj_physical_
1 threads: ------------------------------------------------------
      M.clone()           |      14.3     |
      M.conj_physical_()  |               |          7.6

Times are in microseconds (us).

[-------------- Effective materialization of conj --------------]
                          |  Using copy_  |  Using conj_physical_
1 threads: ------------------------------------------------------
      M.copy_(N)          |      5.9      |
      M.conj_physical_()  |               |          4.4

Times are in microseconds (us).
```
Aliasing checks are certainly far cheaper than always conjugating since we only check if the two tensors in question share storage or not (2 * num_args worst case since we usually don't have more than two mutable argument tensors, and in the worst case all argument tensors are conjugated). 

It's also true that the second approach will simplify the logic (although we'll have to track the mut_arg_clone tensors separately in this case). The perf offset for cloning and copying instead of conjugating in-place twice is 1.7 us and 8-9 us on CPU and CUDA respectively [fyi: the perf difference on CPU is almost invariant of the size]. 

I think the difference on CUDA seems pretty significant to forego the first approach in favor of the latter.

anjali411 replied:
This is a significant perf difference (50% or more) for some ops like `torch.{dot/vdot}` see here: https://github.com/pytorch/pytorch/pull/62915#issuecomment-895338209

ezyang replied:
> This is a significant perf difference (50% or more) for some ops like torch.{dot/vdot} see here: #62915 (comment)

This isn't relevant, because you're going to  bypass MathBitFallback in that case right?

ezyang replied:
I agree that the simplified variant is slower. But everything is slow *anyway* because you're hitting the fallback case (and materializing a bunch of stuff). So the question is how much the perf improvement (in an already slow case) balances with complexity. Well, I suppose that if you really want this optimization, we can keep it, but then I need to rereview the algorithm again.

anjali411 replied:
ok let's get the current version in for 1.10 and we can revisit this after

anjali411 committed:
Update on "Handle shared memory cases in MathBitFallback"


This PR fixes the case when a read and write is performed on a memory shared between mutable and (or) non-mutable arguments. Example:
```
a=torch.tensor([1+1j])
b=a.conj()
b.add_(a) # should return tensor([2]) but returns tensor ([2-2j])
```

The issue here is that in the conjugate fallback, we resolve the conjugation in-place for mutable arguments which can be a problem as shown above in the case when other input arguments share memory with the mutable argument(s). 
This PR fixes this issue by:
1. first scanning through the operator input arguments and creating a vector of mutable arguments that have the conj bit set to `True` (and accordingly setting the flag `check_for_alias_with_mut_arg ` to `True` or `False`).
2. Iterating through all the arguments. At this time we only look at the non-mutable arguments. If `check_for_alias_with_mut_arg` is set to `True`, then we iterate through `mutable_inputs` to check if the current arg tensor in question doesn't alias any of the entries in `mutable_inputs`. If yes, then we clone the non-mutable tensor arg, else we resolve the conjugation as before.
3. Now we look through the mutable_inputs vector (which contains only mutable input tensors with conj bit set to `True`). We in-place conjugate each of the entries in the vector.
4. Do the computation.
5. Re-conjugate the mutable argument tensors.

NOTE: `TensorLists` are not fully handled in ConjugateFallback. Please see the in-line comment for more details.

Fixes https://github.com/pytorch/pytorch/issues/59943

Differential Revision: [D30466905](https://our.internmc.facebook.com/intern/diff/D30466905)

[ghstack-poisoned]

anjali411 committed:
Update on "Handle shared memory cases in MathBitFallback"


This PR fixes the case when a read and write is performed on a memory shared between mutable and (or) non-mutable arguments. Example:
```
a=torch.tensor([1+1j])
b=a.conj()
b.add_(a) # should return tensor([2]) but returns tensor ([2-2j])
```

The issue here is that in the conjugate fallback, we resolve the conjugation in-place for mutable arguments which can be a problem as shown above in the case when other input arguments share memory with the mutable argument(s). 
This PR fixes this issue by:
1. first scanning through the operator input arguments and creating a vector of mutable arguments that have the conj bit set to `True` (and accordingly setting the flag `check_for_alias_with_mut_arg ` to `True` or `False`).
2. Iterating through all the arguments. At this time we only look at the non-mutable arguments. If `check_for_alias_with_mut_arg` is set to `True`, then we iterate through `mutable_inputs` to check if the current arg tensor in question doesn't alias any of the entries in `mutable_inputs`. If yes, then we clone the non-mutable tensor arg, else we resolve the conjugation as before.
3. Now we look through the mutable_inputs vector (which contains only mutable input tensors with conj bit set to `True`). We in-place conjugate each of the entries in the vector.
4. Do the computation.
5. Re-conjugate the mutable argument tensors.

NOTE: `TensorLists` are not fully handled in ConjugateFallback. Please see the in-line comment for more details.

Fixes https://github.com/pytorch/pytorch/issues/59943

Differential Revision: [D30466905](https://our.internmc.facebook.com/intern/diff/D30466905)

[ghstack-poisoned]

anjali411 commented:
@ezyang updated the in-line comments and note based on your suggestion. Also added a perf benchmark in the comment.  unconditionally cloning the mutable inputs would be a big performance hit on CUDA, it seems.

anjali411 commented on aten/src/ATen/native/MathBitsFallback.h:
```diff
@@ -25,25 +25,76 @@ struct MathOpFallback {
+    // set to True if there's one or mutable inputs
     c10::optional<bool> is_write;
+
+    // Mutable inputs to be tracked separately
```
```suggestion
    // Conjugated Mutable inputs to be tracked separately
```

anjali411 commented on aten/src/ATen/native/MathBitsFallback.h:
```diff
@@ -25,25 +25,76 @@ struct MathOpFallback {
     c10::optional<bool> is_write;
+
+    // Mutable inputs to be tracked separately
+    std::vector<Tensor> mutable_inputs;
```
```suggestion
    std::vector<Tensor> conjugated_mutable_inputs;
```

albanD commented on aten/src/ATen/native/MathBitsFallback.h:
```diff
@@ -72,55 +134,61 @@ struct MathOpFallback {
+            // might be sharing memory with one of the non-mutable tensors that could be reading into the wrong
+            // values if their math bit is not set to True.
+            // no op if tensor and mutable arg share memory and have math bit set to True
+            if (tensor.is_alias_of(mutable_input)) {
```
What is this function checking for? Full aliasing like they look at the exactly the same memory but not a subset?
Partial overlaps will just get ignored here and materialized out of place because resolve_needed remains "true"?

anjali411 replied:
synced offline: check for memory overlap using https://github.com/pytorch/pytorch/blob/839eaa2e91556ecd4532596b4fef18a1c3f6e1c1/aten/src/ATen/MemoryOverlap.cpp#L39-L72

anjali411 commented on aten/src/ATen/native/MathBitsFallback.h:
```diff
@@ -60,6 +111,17 @@ struct MathOpFallback {
         } else {
           is_write = alias_info->isWrite();
+          auto& ivalue = (*stack)[stack_start + i];
+          //TODO: add a table listing all possible cases and clearly state what works/is allowed
```
yeah forgot to remove it

anjali411 commented on aten/src/ATen/native/MathBitsFallback.h:
```diff
@@ -25,25 +25,76 @@ struct MathOpFallback {
+        WRITE from an argument.  Conservative approach is to assume that we always
+        READ from an argument, but in out-of-place operations you can skip
+        conjugating inputs on entry that never get used.  In current schema we
+        can't easily tell if inplace situation has happened, so don't do it.
```
updated the description

anjali411 committed:
Update on "Handle shared memory cases in MathBitFallback"


This PR fixes the case when a read and write is performed on a memory shared between mutable and (or) non-mutable arguments. Example:
```
a=torch.tensor([1+1j])
b=a.conj()
b.add_(a) # should return tensor([2]) but returns tensor ([2-2j])
```

The issue here is that in the conjugate fallback, we resolve the conjugation in-place for mutable arguments which can be a problem as shown above in the case when other input arguments share memory with the mutable argument(s). 
This PR fixes this issue by:
1. first scanning through the operator input arguments and creating a vector of mutable arguments that have the conj bit set to `True` (and accordingly setting the flag `check_for_alias_with_mut_arg ` to `True` or `False`).
2. Iterating through all the arguments. At this time we only look at the non-mutable arguments. If `check_for_alias_with_mut_arg` is set to `True`, then we iterate through `mutable_inputs` to check if the current arg tensor in question doesn't alias any of the entries in `mutable_inputs`. If yes, then we clone the non-mutable tensor arg, else we resolve the conjugation as before.
3. Now we look through the mutable_inputs vector (which contains only mutable input tensors with conj bit set to `True`). We in-place conjugate each of the entries in the vector.
4. Do the computation.
5. Re-conjugate the mutable argument tensors.

NOTE: `TensorLists` are not fully handled in ConjugateFallback. Please see the in-line comment for more details.

Fixes https://github.com/pytorch/pytorch/issues/59943

Differential Revision: [D30466905](https://our.internmc.facebook.com/intern/diff/D30466905)

[ghstack-poisoned]

anjali411 committed:
Update on "Handle shared memory cases in MathBitFallback"


This PR fixes the case when a read and write is performed on a memory shared between mutable and (or) non-mutable arguments. Example:
```
a=torch.tensor([1+1j])
b=a.conj()
b.add_(a) # should return tensor([2]) but returns tensor ([2-2j])
```

The issue here is that in the conjugate fallback, we resolve the conjugation in-place for mutable arguments which can be a problem as shown above in the case when other input arguments share memory with the mutable argument(s). 
This PR fixes this issue by:
1. first scanning through the operator input arguments and creating a vector of mutable arguments that have the conj bit set to `True` (and accordingly setting the flag `check_for_alias_with_mut_arg ` to `True` or `False`).
2. Iterating through all the arguments. At this time we only look at the non-mutable arguments. If `check_for_alias_with_mut_arg` is set to `True`, then we iterate through `mutable_inputs` to check if the current arg tensor in question doesn't alias any of the entries in `mutable_inputs`. If yes, then we clone the non-mutable tensor arg, else we resolve the conjugation as before.
3. Now we look through the mutable_inputs vector (which contains only mutable input tensors with conj bit set to `True`). We in-place conjugate each of the entries in the vector.
4. Do the computation.
5. Re-conjugate the mutable argument tensors.

NOTE: `TensorLists` are not fully handled in ConjugateFallback. Please see the in-line comment for more details.

Fixes https://github.com/pytorch/pytorch/issues/59943

Differential Revision: [D30466905](https://our.internmc.facebook.com/intern/diff/D30466905)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30466905).

malfet commented:
Tentatively moving to 1.11

ezyang commented:
Mm, we should still land (a simplified version) of this :)

anjali411 committed:
Update on "Handle shared memory cases in MathBitFallback"


This PR fixes the case when a read and write is performed on a memory shared between mutable and (or) non-mutable arguments. Example:
```
a=torch.tensor([1+1j])
b=a.conj()
b.add_(a) # should return tensor([2]) but returns tensor ([2-2j])
```

The issue here is that in the conjugate fallback, we resolve the conjugation in-place for mutable arguments which can be a problem as shown above in the case when other input arguments share memory with the mutable argument(s). 
This PR fixes this issue by:
1. first scanning through the operator input arguments and creating a vector of mutable arguments that have the conj bit set to `True` (and accordingly setting the flag `check_for_alias_with_mut_arg ` to `True` or `False`).
2. Iterating through all the arguments. At this time we only look at the non-mutable arguments. If `check_for_alias_with_mut_arg` is set to `True`, then we iterate through `mutable_inputs` to check if the current arg tensor in question doesn't alias any of the entries in `mutable_inputs`. If yes, then we clone the non-mutable tensor arg, else we resolve the conjugation as before.
3. Now we look through the mutable_inputs vector (which contains only mutable input tensors with conj bit set to `True`). We in-place conjugate each of the entries in the vector.
4. Do the computation.
5. Re-conjugate the mutable argument tensors.

NOTE: `TensorLists` are not fully handled in ConjugateFallback. Please see the in-line comment for more details.

Fixes https://github.com/pytorch/pytorch/issues/59943

Differential Revision: [D30466905](https://our.internmc.facebook.com/intern/diff/D30466905)

[ghstack-poisoned]

anjali411 committed:
Update on "Handle shared memory cases in MathBitFallback"


This PR fixes the case when a read and write is performed on a memory shared between mutable and (or) non-mutable arguments. Example:
```
a=torch.tensor([1+1j])
b=a.conj()
b.add_(a) # should return tensor([2]) but returns tensor ([2-2j])
```

The issue here is that in the conjugate fallback, we resolve the conjugation in-place for mutable arguments which can be a problem as shown above in the case when other input arguments share memory with the mutable argument(s). 
This PR fixes this issue by:
1. first scanning through the operator input arguments and creating a vector of mutable arguments that have the conj bit set to `True` (and accordingly setting the flag `check_for_alias_with_mut_arg ` to `True` or `False`).
2. Iterating through all the arguments. At this time we only look at the non-mutable arguments. If `check_for_alias_with_mut_arg` is set to `True`, then we iterate through `mutable_inputs` to check if the current arg tensor in question doesn't alias any of the entries in `mutable_inputs`. If yes, then we clone the non-mutable tensor arg, else we resolve the conjugation as before.
3. Now we look through the mutable_inputs vector (which contains only mutable input tensors with conj bit set to `True`). We in-place conjugate each of the entries in the vector.
4. Do the computation.
5. Re-conjugate the mutable argument tensors.

NOTE: `TensorLists` are not fully handled in ConjugateFallback. Please see the in-line comment for more details.

Fixes https://github.com/pytorch/pytorch/issues/59943

Differential Revision: [D30466905](https://our.internmc.facebook.com/intern/diff/D30466905)

cc ezyang anjali411 dylanbespalko mruberry @Lezcano nikitaved

[ghstack-poisoned]

anjali411 committed:
Update on "Handle shared memory cases in MathBitFallback"


This PR fixes the case when a read and write is performed on a memory shared between mutable and (or) non-mutable arguments. Example:
```
a=torch.tensor([1+1j])
b=a.conj()
b.add_(a) # should return tensor([2]) but returns tensor ([2-2j])
```

The issue here is that in the conjugate fallback, we resolve the conjugation in-place for mutable arguments which can be a problem as shown above in the case when other input arguments share memory with the mutable argument(s). 
This PR fixes this issue by:
1. first scanning through the operator input arguments and creating a vector of mutable arguments that have the conj bit set to `True` (and accordingly setting the flag `check_for_alias_with_mut_arg ` to `True` or `False`).
2. Iterating through all the arguments. At this time we only look at the non-mutable arguments. If `check_for_alias_with_mut_arg` is set to `True`, then we iterate through `mutable_inputs` to check if the current arg tensor in question doesn't alias any of the entries in `mutable_inputs`. If yes, then we clone the non-mutable tensor arg, else we resolve the conjugation as before.
3. Now we look through the mutable_inputs vector (which contains only mutable input tensors with conj bit set to `True`). We in-place conjugate each of the entries in the vector.
4. Do the computation.
5. Re-conjugate the mutable argument tensors.

NOTE: `TensorLists` are not fully handled in ConjugateFallback. Please see the in-line comment for more details.

Fixes https://github.com/pytorch/pytorch/issues/59943

Differential Revision: [D30466905](https://our.internmc.facebook.com/intern/diff/D30466905)

cc ezyang anjali411 dylanbespalko mruberry @Lezcano nikitaved

[ghstack-poisoned]

anjali411 committed:
Update on "Handle shared memory cases in MathBitFallback"


This PR fixes the case when a read and write is performed on a memory shared between mutable and (or) non-mutable arguments. Example:
```
a=torch.tensor([1+1j])
b=a.conj()
b.add_(a) # should return tensor([2]) but returns tensor ([2-2j])
```

The issue here is that in the conjugate fallback, we resolve the conjugation in-place for mutable arguments which can be a problem as shown above in the case when other input arguments share memory with the mutable argument(s). 
This PR fixes this issue by:
1. first scanning through the operator input arguments and creating a vector of mutable arguments that have the conj bit set to `True` (and accordingly setting the flag `check_for_alias_with_mut_arg ` to `True` or `False`).
2. Iterating through all the arguments. At this time we only look at the non-mutable arguments. If `check_for_alias_with_mut_arg` is set to `True`, then we iterate through `mutable_inputs` to check if the current arg tensor in question doesn't alias any of the entries in `mutable_inputs`. If yes, then we clone the non-mutable tensor arg, else we resolve the conjugation as before.
3. Now we look through the mutable_inputs vector (which contains only mutable input tensors with conj bit set to `True`). We in-place conjugate each of the entries in the vector.
4. Do the computation.
5. Re-conjugate the mutable argument tensors.

NOTE: `TensorLists` are not fully handled in ConjugateFallback. Please see the in-line comment for more details.

Fixes https://github.com/pytorch/pytorch/issues/59943

Differential Revision: [D30466905](https://our.internmc.facebook.com/intern/diff/D30466905)

cc ezyang anjali411 dylanbespalko mruberry @Lezcano nikitaved

[ghstack-poisoned]

anjali411 committed:
Update on "Handle shared memory cases in MathBitFallback"


This PR fixes the case when a read and write is performed on a memory shared between mutable and (or) non-mutable arguments. Example:
```
a=torch.tensor([1+1j])
b=a.conj()
b.add_(a) # should return tensor([2]) but returns tensor ([2-2j])
```

The issue here is that in the conjugate fallback, we resolve the conjugation in-place for mutable arguments which can be a problem as shown above in the case when other input arguments share memory with the mutable argument(s). 
This PR fixes this issue by:
1. first scanning through the operator input arguments and creating a vector of mutable arguments that have the conj bit set to `True` (and accordingly setting the flag `check_for_alias_with_mut_arg ` to `True` or `False`).
2. Iterating through all the arguments. At this time we only look at the non-mutable arguments. If `check_for_alias_with_mut_arg` is set to `True`, then we iterate through `mutable_inputs` to check if the current arg tensor in question doesn't alias any of the entries in `mutable_inputs`. If yes, then we clone the non-mutable tensor arg, else we resolve the conjugation as before.
3. Now we look through the mutable_inputs vector (which contains only mutable input tensors with conj bit set to `True`). We in-place conjugate each of the entries in the vector.
4. Do the computation.
5. Re-conjugate the mutable argument tensors.

NOTE: `TensorLists` are not fully handled in ConjugateFallback. Please see the in-line comment for more details.

Fixes https://github.com/pytorch/pytorch/issues/59943

Differential Revision: [D30466905](https://our.internmc.facebook.com/intern/diff/D30466905)

cc ezyang anjali411 dylanbespalko mruberry @Lezcano nikitaved

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30466905).

anjali411 committed:
Update on "Handle shared memory cases in MathBitFallback"


This PR fixes the case when a read and write is performed on a memory shared between mutable and (or) non-mutable arguments. Example:
```
a=torch.tensor([1+1j])
b=a.conj()
b.add_(a) # should return tensor([2]) but returns tensor ([2-2j])
```

The issue here is that in the conjugate fallback, we resolve the conjugation in-place for mutable arguments which can be a problem as shown above in the case when other input arguments share memory with the mutable argument(s). 
This PR fixes this issue by:
1. first scanning through the operator input arguments and creating a vector of mutable arguments that have the conj bit set to `True` (and accordingly setting the flag `check_for_alias_with_mut_arg ` to `True` or `False`).
2. Iterating through all the arguments. At this time we only look at the non-mutable arguments. If `check_for_alias_with_mut_arg` is set to `True`, then we iterate through `mutable_inputs` to check if the current arg tensor in question doesn't alias any of the entries in `mutable_inputs`. If yes, then we clone the non-mutable tensor arg, else we resolve the conjugation as before.
3. Now we look through the mutable_inputs vector (which contains only mutable input tensors with conj bit set to `True`). We in-place conjugate each of the entries in the vector.
4. Do the computation.
5. Re-conjugate the mutable argument tensors.

NOTE: `TensorLists` are not fully handled in ConjugateFallback. Please see the in-line comment for more details.

Fixes https://github.com/pytorch/pytorch/issues/59943

Differential Revision: [D30466905](https://our.internmc.facebook.com/intern/diff/D30466905)

cc ezyang anjali411 dylanbespalko mruberry @Lezcano nikitaved

[ghstack-poisoned]

anjali411 committed:
Update on "Handle shared memory cases in MathBitFallback"


This PR fixes the case when a read and write is performed on a memory shared between mutable and (or) non-mutable arguments. Example:
```
a=torch.tensor([1+1j])
b=a.conj()
b.add_(a) # should return tensor([2]) but returns tensor ([2-2j])
```

The issue here is that in the conjugate fallback, we resolve the conjugation in-place for mutable arguments which can be a problem as shown above in the case when other input arguments share memory with the mutable argument(s). 
This PR fixes this issue by:
1. first scanning through the operator input arguments and creating a vector of mutable arguments that have the conj bit set to `True` (and accordingly setting the flag `check_for_alias_with_mut_arg ` to `True` or `False`).
2. Iterating through all the arguments. At this time we only look at the non-mutable arguments. If `check_for_alias_with_mut_arg` is set to `True`, then we iterate through `mutable_inputs` to check if the current arg tensor in question doesn't alias any of the entries in `mutable_inputs`. If yes, then we clone the non-mutable tensor arg, else we resolve the conjugation as before.
3. Now we look through the mutable_inputs vector (which contains only mutable input tensors with conj bit set to `True`). We in-place conjugate each of the entries in the vector.
4. Do the computation.
5. Re-conjugate the mutable argument tensors.

NOTE: `TensorLists` are not fully handled in ConjugateFallback. Please see the in-line comment for more details.

Fixes https://github.com/pytorch/pytorch/issues/59943

Differential Revision: [D30466905](https://our.internmc.facebook.com/intern/diff/D30466905)

cc ezyang anjali411 dylanbespalko mruberry @Lezcano nikitaved

[ghstack-poisoned]

anjali411 committed:
Update on "Handle shared memory cases in MathBitFallback"


This PR fixes the case when a read and write is performed on a memory shared between mutable and (or) non-mutable arguments. Example:
```
a=torch.tensor([1+1j])
b=a.conj()
b.add_(a) # should return tensor([2]) but returns tensor ([2-2j])
```

The issue here is that in the conjugate fallback, we resolve the conjugation in-place for mutable arguments which can be a problem as shown above in the case when other input arguments share memory with the mutable argument(s). 
This PR fixes this issue by:
1. first scanning through the operator input arguments and creating a vector of mutable arguments that have the conj bit set to `True` (and accordingly setting the flag `check_for_alias_with_mut_arg ` to `True` or `False`).
2. Iterating through all the arguments. At this time we only look at the non-mutable arguments. If `check_for_alias_with_mut_arg` is set to `True`, then we iterate through `mutable_inputs` to check if the current arg tensor in question doesn't alias any of the entries in `mutable_inputs`. If yes, then we clone the non-mutable tensor arg, else we resolve the conjugation as before.
3. Now we look through the mutable_inputs vector (which contains only mutable input tensors with conj bit set to `True`). We in-place conjugate each of the entries in the vector.
4. Do the computation.
5. Re-conjugate the mutable argument tensors.

NOTE: `TensorLists` are not fully handled in ConjugateFallback. Please see the in-line comment for more details.

Fixes https://github.com/pytorch/pytorch/issues/59943

Differential Revision: [D30466905](https://our.internmc.facebook.com/intern/diff/D30466905)

cc ezyang anjali411 dylanbespalko mruberry  nikitaved

[ghstack-poisoned]

anjali411 committed:
Update on "Handle shared memory cases in MathBitFallback"


This PR fixes the case when a read and write is performed on a memory shared between mutable and (or) non-mutable arguments. Example:
```
a=torch.tensor([1+1j])
b=a.conj()
b.add_(a) # should return tensor([2]) but returns tensor ([2-2j])
```

The issue here is that in the conjugate fallback, we resolve the conjugation in-place for mutable arguments which can be a problem as shown above in the case when other input arguments share memory with the mutable argument(s). 
This PR fixes this issue by:
1. first scanning through the operator input arguments and creating a vector of mutable arguments that have the conj bit set to `True` (and accordingly setting the flag `check_for_alias_with_mut_arg ` to `True` or `False`).
2. Iterating through all the arguments. At this time we only look at the non-mutable arguments. If `check_for_alias_with_mut_arg` is set to `True`, then we iterate through `mutable_inputs` to check if the current arg tensor in question doesn't alias any of the entries in `mutable_inputs`. If yes, then we clone the non-mutable tensor arg, else we resolve the conjugation as before.
3. Now we look through the mutable_inputs vector (which contains only mutable input tensors with conj bit set to `True`). We in-place conjugate each of the entries in the vector.
4. Do the computation.
5. Re-conjugate the mutable argument tensors.

NOTE: `TensorLists` are not fully handled in ConjugateFallback. Please see the in-line comment for more details.

Fixes https://github.com/pytorch/pytorch/issues/59943

Differential Revision: [D30466905](https://our.internmc.facebook.com/intern/diff/D30466905)

cc ezyang anjali411 dylanbespalko mruberry  nikitaved

[ghstack-poisoned]

anjali411 committed:
Update on "Handle shared memory cases in MathBitFallback"


This PR fixes the case when a read and write is performed on a memory shared between mutable and (or) non-mutable arguments. Example:
```
a=torch.tensor([1+1j])
b=a.conj()
b.add_(a) # should return tensor([2]) but returns tensor ([2-2j])
```

The issue here is that in the conjugate fallback, we resolve the conjugation in-place for mutable arguments which can be a problem as shown above in the case when other input arguments share memory with the mutable argument(s). 
This PR fixes this issue by:
1. first scanning through the operator input arguments and creating a vector of mutable arguments that have the conj bit set to `True` (and accordingly setting the flag `check_for_alias_with_mut_arg ` to `True` or `False`).
2. Iterating through all the arguments. At this time we only look at the non-mutable arguments. If `check_for_alias_with_mut_arg` is set to `True`, then we iterate through `mutable_inputs` to check if the current arg tensor in question doesn't alias any of the entries in `mutable_inputs`. If yes, then we clone the non-mutable tensor arg, else we resolve the conjugation as before.
3. Now we look through the mutable_inputs vector (which contains only mutable input tensors with conj bit set to `True`). We in-place conjugate each of the entries in the vector.
4. Do the computation.
5. Re-conjugate the mutable argument tensors.

NOTE: `TensorLists` are not fully handled in ConjugateFallback. Please see the in-line comment for more details.

Fixes https://github.com/pytorch/pytorch/issues/59943

Differential Revision: [D30466905](https://our.internmc.facebook.com/intern/diff/D30466905)

cc ezyang anjali411 dylanbespalko mruberry  nikitaved

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30466905).
